proposals/nnnn-freestanding-macros.md

```diff
@@ -237,26 +237,77 @@ public struct Int64 { ... }
 public struct UInt64 { ... }
 ```
 
-### Short hand for `@main`
+### Template code generator
 
-A freestanding declaration macro `#main` could be defined as a short hand for a `@main` struct with a `main()` method.
+The Swift Standard Library makes extensive use of the [gyb](https://github.com/apple/swift/blob/main/utils/gyb.py) tool to generate boilerplate-y Swift code such as [tgmath.swift.gyb](https://github.com/apple/swift/blob/main/stdlib/public/Platform/tgmath.swift.gyb). The template code is written in `.gyb` files, which are processed by the gyb tool separately before Swift compilation. With freestanding declaration macros, one could write a macro to accept a string as a template and a list of replacement values, allowing templates to be defined inline and eliminating the need to set up a separate build phase.
 
 ```swift
-@freestanding(declaration)
-macro main(_ body: () -> Void)
+@freestanding(declaration, names: arbitrary)
+macro gyb(String, [Any])
+
+#gyb({
+  public struct Int${0} { ... }
+  public struct UInt${0} { ... }
+}, [8, 16, 32, 64])
+```
+
+This expands to:
+
+```swift
+  public struct Int8 { ... }
+  public struct UInt8 { ... }
+
+  public struct Int16 { ... }
+  public struct UInt16 { ... }
+
+  public struct Int32 { ... }
+  public struct UInt32 { ... }
 
-#main {
-  print("Hello")
+  public struct Int64 { ... }
+  public struct UInt64 { ... }
+```
+
+### Data model generator
+
+Declaring a data model for an existing textual serialization may need some amount of eyeballing and is prone to errors.  A freestanding declaration macro can be used to analyze a template textual serialization, e.g. JSON, and declare a model data structure against the template.
+
+```swift
+@freestanding(declaration, names: arbitrary)
+macro jsonModel(String)
+
+struct JSONValue: Codable {
+  #jsonModel("""
```

Comment:
Here I intentionally had the `jsonModel` macro _not_ produce a top-level model type, because the user should be able to control its access level and its name.

This may arguably be better as an attached macro, where you attach `@jsonModel` with a string on top of a struct declaration, but I do appreciate being able to write the template _inside_ a struct, not on top of it. So I think it's a good declaration macro example.